### PR TITLE
QuickStart Actions

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Addresses
https://issues.redhat.com/browse/ODC-5659

# Screenshots
Screencast from 04-14-2021 07_03_37 PM
Screencast from 04-14-2021 07_04_16 PM
Screencast from 04-15-2021 10_36_27 PM

# Tests
Added suite for QuickStartFooter

# Browser Conformance
Chrome

### CODE REVIEW

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 1]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
-    <div className={`co-quick-start-footer ${footerClass}`}>
+  const getPrimaryButton = (
+    <Button
+      style={{
+        marginRight: 'var(--pf-global--spacer--md)',
+      }}
+      type="submit"
+      variant="primary"
+      onClick={onNext}
+      isInline
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:75]:
```suggestion
```

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 2]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
+  };
 
-  return (
-    <div className={`co-quick-start-footer ${footerClass}`}>
+  const getPrimaryButton = (
+    <Button
+      style={{
+        marginRight: 'var(--pf-global--spacer--md)',
+      }}
+      type="submit"
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:72]:
```suggestion
```

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 3]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
       <Button
         style={{
           marginRight: 'var(--pf-global--spacer--md)',
         }}
         type="submit"
-        variant="primary"
-        onClick={onNext}
+        variant="secondary"
+        onClick={onRestart}
         isInline
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:90]:
```suggestion
```

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 4]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
+    </Button>
+  );
+
+  const getSecondaryButton =
+    taskNumber === -1 && status === QuickStartStatus.IN_PROGRESS ? (
       <Button
         style={{
           marginRight: 'var(--pf-global--spacer--md)',
         }}
         type="submit"
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:87]:
```suggestion
```

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 5]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
   const quickStartPath = '/quickstart';
   const { t } = useTranslation();
 
+  const { restartQuickStart } = React.useContext<QuickStartContextValues>(QuickStartContext);
+
   const PrimaryButtonText = {
-    START: t('quickstart~Start tour'),
+    START: t('quickstart~Start'),
+    CONTINUE: t('quickstart~Continue'),
     NEXT: t('quickstart~Next'),
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:39]: use `useMemo`

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 6]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
     CLOSE: t('quickstart~Close'),
   };
 
-  const getPrimaryButtonText = React.useCallback((): string => {
+  const SecondaryButtonText = {
+    BACK: t('quickstart~Back'),
+    RESTART: t('quickstart~Restart'),
+  };
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:46]: `useMemo`

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 7]
```diff
@@ -22,46 +24,75 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
-  return (
-    <div className={`co-quick-start-footer ${footerClass}`}>
+  const getPrimaryButton = (
+    <Button
+      style={{
+        marginRight: 'var(--pf-global--spacer--md)',
+      }}
+      type="submit"
+      variant="primary"
+      onClick={onNext}
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:74]: If pressed on `Continue` button, should go the next unstarted step

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 8]
```diff
@@ -71,14 +102,36 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
+        </Button>
+      )
+    );
+
+  const getSideNoteAction =
+    status === QuickStartStatus.COMPLETE &&
+    (currentPath !== quickStartPath ? (
+      <Link style={{ display: 'inline-block' }} to={quickStartPath}>
+        {t('quickstart~View all tours')}
+      </Link>
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:115]: When does this link appear on Quick start?

**Review by sahil143 — CHANGES_REQUESTED** [review]
@abhinandan13jan If the user clicks on the `continue` button, it should take user to next unstarted step of quick start. it currently takes user to 1st step.

continue-quick-start

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 8 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: when /quickstart is not the URL like, quicksearch

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 7 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: updated

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 6 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:46]: useMemo added

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 4 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: removed submit

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 3 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: removed inLine

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 9]
```diff
@@ -71,14 +102,36 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
...
+    ) : (
+      taskNumber === totalTasks && (
+        <Button
+          type="submit"
+          variant="link"
+          className="co-quick-start-footer__restartlink pull-right"
+          onClick={onRestart}
+          isInline
+        >
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: updated

**Comment by abhinandan13jan** [comment]
> @abhinandan13jan If the user clicks on the `continue` button, it should take user to next unstarted step of quick start. it currently takes user to 1st step.
> 
> continue-quick-start



@sahil143 updated

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 8 above)
Comment (christianvogt) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: Do we know when / where this requirement came from (adding a `view all tours` link at the end of a quick start)
cc @bmignano @serenamarie125 

This should read `View all quick starts` if we keep it.

Comment (bmignano) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: I think we had that link in the very first iteration of quick starts (apparently back when they were still being called tours 😄) . It doesn't seem like it ever got in, but honestly I don't mind not having it. Thoughts @serenamarie125 ?

File: frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts [hunk 10]
```diff
@@ -187,6 +188,23 @@ export const useValuesForQuickStartContext = (): QuickStartContextValues => {
     [setAllQuickStartStates],
   );
 
+  const getQuickStartActiveTask = useCallback(
+    (totalTasks) => {
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:192]: What is this totalTasks?

File: frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts [hunk 11]
```diff
@@ -187,6 +188,23 @@ export const useValuesForQuickStartContext = (): QuickStartContextValues => {
...
+      let activeTaskNumber = 0;
+      setAllQuickStartStates((qs) => {
+        const quickStart = qs[activeQuickStartID];
+        while (
+          activeTaskNumber !== totalTasks &&
+          quickStart[`taskStatus${activeTaskNumber}`] !== QuickStartTaskStatus.INIT
+        ) {
+          activeTaskNumber++;
+        }
+      });
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:202]: Why are you calling the setState here and you are not returning a new state as well?
looks like you just want to loop over All the quickstarts and find the active taskNumber for an active quick start.
You can just use the activeQuickStartState for that.

File: frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts [hunk 12]
```diff
@@ -187,6 +188,23 @@ export const useValuesForQuickStartContext = (): QuickStartContextValues => {
     [setAllQuickStartStates],
   );
 
+  const getQuickStartActiveTask = useCallback(
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:191]: I don't think you need this method at all.

you can just use the activeQuickStartState in the component that it is being used to get the activeTaskNumber for the activeQuickStart

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx (hunk 8 above)
Comment (serenamarie125) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx]: Please remove the _View all tours_ link!

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts (hunk 12 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts]: moved it to QuickStartController and computing activeTask inside controller

**Review by sahil143 — APPROVED** [review]
/lgtm

verified changes locally

**Pull request merged**